events/api.py

```diff
@@ -2484,15 +2517,76 @@ def perform_update(self, serializer):
 
     @atomic
     def bulk_update(self, request, *args, **kwargs):
-        return super().bulk_update(request, *args, **kwargs)
+        context = self.get_serializer_context()
+
+        context['keywords'], context['location'], bulk = self.cache_related_fields(request)
+        serializer = EventSerializer(self.filter_queryset(self.get_queryset()),
+                                     data=request.data,
+                                     context=context,
+                                     many=bulk,
+                                     partial=partial,)
+        if not serializer.instance:
+            raise serializers.ValidationError('Missing matching events to update.')
+        serializer.is_valid(raise_exception=True)
+        self.perform_update(serializer)
+        return Response(serializer.data, status=status.HTTP_200_OK)
 
     @atomic
     def create(self, request, *args, **kwargs):
-        return super().create(request, *args, **kwargs)
+        context = self.get_serializer_context()
+        context['keywords'], context['location'], bulk = self.cache_related_fields(request)
+        serializer = EventSerializer(data=request.data,
+                                     context=context,
+                                     many=bulk,)
+        serializer.is_valid(raise_exception=True)
+        self.perform_create(serializer)
+        headers = self.get_success_headers(serializer.data)
+        return Response(serializer.data, status=status.HTTP_201_CREATED, headers=headers)
+
+    def cache_related_fields(self, request):
+
+        def retrieve_ids(key, event):
+            if key in event.keys():
+                if isinstance(event[key], list):
+                    event_list = event.get(key, [])
+                else:
+                    event_list = [event.get(key, [])]
+                ids = [i['@id'].rstrip('/').split('/').pop() for i in event_list if i and
+                                                                                    isinstance(i, dict) and  # noqa E127 
+                                                                                    i.get('@id', None)]  # noqa E127
+                return ids
+            else:
+                return []
+
+        bulk = isinstance(request.data, list)
+
+        if not bulk:
+            events = [request.data]
+        else:
+            events = request.data
+
+        keywords = Keyword.objects.none()
+        locations = Place.objects.none()
+        keyword_ids = []
+        location_ids = []
+        for event in events:
+            keyword_ids.extend(retrieve_ids('keywords', event))
+            keyword_ids.extend(retrieve_ids('audience', event))
+            location_ids.extend(retrieve_ids('location', event))
+        if location_ids:
+            locations = Place.objects.filter(id__in=location_ids)
+        if keyword_ids:
+            keywords = Keyword.objects.filter(id__in=keyword_ids)
+        return keywords, locations, bulk
```

Comment:
Could the retrieve ids be a more common util? It seems like it could be handy in some other parts of the code also. Changes in cache related fields seems to 4th subject in same commit.